### PR TITLE
Update Composite spaces with full coverage

### PR DESCRIPTION
**Pull request by pseudo-rnd-thoughts**
Update all of the composite spaces (`Dict`, `Tuple`, `Graph` and `Sequence`) with full coverage
This is the first part of #2977 split into two

These tests now provide 100% coverage for all of the composite spaces

Changes
1. Allows both `spaces` and `kwargs` to be used in `Dict` with checks that the same keys are not in both 
2. Simplifies `contains` in `Dict` using positive (not negative) checks
3. When setting an element in a `Dict`, checks that the space is a gym space
4. Simplifies `Dict.from_jsonable` to use list comprehension and returns `OrderDict` such that roundtripping a sample is True
5. Updated `GraphInstance` to use class style with attribute
6. Update type hints in `Graph`
7. Change `Graph.contains` to use positive (not negative) checks
8. Update `Sequence.sample` to use `int` and `np.ndarray` masks with assertation checks
9. Update `Tuple.seed` type hints and assertations
10. Update `_graph_unflatten` arguments to not conflict with `_flatten_graph`
11. Remove composite space tests from `test_spaces` and add to space specify testing files. Adds a number of new tests (a couple are for coverage and checking the errors occur for expected inputs). 

A second PR will add similar updates for fundamental spaces and add generalised tests for all spaces including composite spaces

### CODE REVIEW

**Review by arjun-kg — COMMENTED** [review]
Checked a few files, left some comments about additional useful tests. I've commented on Dict and Tuple spaces, but they might be applicable on all the spaces. 

Edit 1: 
Mentioning again here - Sequence space allows 0 length sequences, but `space.sample()` can never sample 0 length sequences since it comes from `np.random.geometric`. A quick fix can be `np.random.geometric(0.25) - 1`.

File: tests/spaces/test_dict.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        dict_sample = DICT_SPACE.sample()
+        assert np.all(dict_sample["a"] == a.sample())
+        assert np.all(dict_sample["b"]["b_1"] == b_1.sample())
+        assert np.all(dict_sample["b"]["b_2"] == b_2.sample())
+        assert dict_sample["c"] == c.sample()
+
+
+def test_none_seeding():
+    seeds = DICT_SPACE.seed(None)
+    assert len(seeds) == 4 and all(isinstance(seed, int) for seed in seeds)
```
[the remark this answers is not in the document]
Comment (arjun-kg) [reply tests/spaces/test_dict.py:108]: Or even directly comparing two samples

[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply tests/spaces/test_dict.py:108]: The problem is that while the chances of the same random key occurring are nearly zero, there is still a chance so I don't think this matter.

File: tests/spaces/test_tuple.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    tuple_space = Tuple(spaces)
+
+    assert len(tuple_space) == len(spaces)
+    # Test indexing
+    for i in range(len(tuple_space)):
+        assert tuple_space[i] == spaces[i]
+
+    # Test iterable
+    for space in tuple_space:
+        assert space in spaces
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply tests/spaces/test_tuple.py:20]: As they are tuples then there are naturally ordered so we shouldn't need to test to my knowledge

File: tests/spaces/test_tuple.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    # Test count
+    assert tuple_space.count(Discrete(5)) == 2
+    assert tuple_space.count(MultiBinary(2)) == 0
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply tests/spaces/test_tuple.py:24]: Added

File: tests/spaces/test_tuple.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            (123, {"position": 456, "velocity": 789}),
+            3,
+        ),
+    ],
+)
+def test_seeds(space, seed, expected_len):
+    seeds = space.seed(seed)
+    assert isinstance(seeds, list) and all(isinstance(elem, int) for elem in seeds)
+    assert len(seeds) == expected_len
+
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply tests/spaces/test_tuple.py:66]: Added

File: tests/spaces/test_dict.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+    DICT_SPACE["a"] = a
+
+
+def test_iterator():
+    """Tests the Dict `__iter__` function correctly returns keys in the subspaces"""
+    for key in DICT_SPACE:
+        assert key in DICT_SPACE.spaces
+
+    assert {key for key in DICT_SPACE} == DICT_SPACE.spaces.keys()
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply tests/spaces/test_dict.py:140]: yes and no, in the #3048 such testing is added and this PR will be merged first

File: gym/spaces/sequence.py [hunk 6]
```diff
@@ -50,29 +54,49 @@ def is_np_flattenable(self):
...
-            length_mask = None
-            feature_mask = None
+            length_mask, feature_mask = None, None
+
         if length_mask is not None:
-            length = self.np_random.choice(length_mask)
+            if np.issubdtype(type(length_mask), np.integer):
+                assert (
+                    0 <= length_mask
+                ), f"Expects the length mask to be greater than zero, actual value: {length_mask}"
```
Comment (arjun-kg) [inline gym/spaces/sequence.py:84]: I guess the text should say greater than or equal to zero if we're allowing zero length sequences.

File: gym/spaces/sequence.py [hunk 7]
```diff
@@ -50,29 +54,49 @@ def is_np_flattenable(self):
...
+                length = length_mask
+            elif isinstance(length_mask, np.ndarray):
+                assert (
+                    len(length_mask.shape) == 1
+                ), f"Expects the shape of the length mask to be 1-dimensional, actual shape: {length_mask.shape}"
+                assert np.all(
+                    0 <= length_mask
+                ), f"Expects all values in the length_mask to be greater than zero, actual values: {length_mask}"
```
Comment (arjun-kg) [inline gym/spaces/sequence.py:92]: Same with including zero

File: gym/spaces/tuple.py [hunk 8]
```diff
@@ -57,25 +63,32 @@ def seed(self, seed: Optional[Union[int, List[int]]] = None) -> list:
         """
         seeds = []
 
-        if isinstance(seed, list):
-            for i, space in enumerate(self.spaces):
-                seeds += space.seed(seed[i])
+        if isinstance(seed, Sequence):
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gym/spaces/tuple.py:66]: Agreed, changed

**Pull request merged**